qcp/sudoku.py

```diff
@@ -0,0 +1,169 @@
+        cond = cond4 * cond3 * cond2 * cond1
+        return cond
+
+    def diffusion(self):
+        """
+        Creates a diffusion gate - a gate which amplifies the probability of
+        selecting our target state
+
+        returns:
+            Matrix: Matrix representing diffusion gate
+        """
+        had = g.multi_gate(9, [0, 1, 2, 3], g.Gate.H)
+        xs = g.multi_gate(9, [0, 1, 2, 3], g.Gate.X)
+        cz = g.control_z(9, [0, 1, 2], 3)
+
+        diff = had * xs * cz * xs * had
+        return diff
+
+    def construct_circuit(self):
+        """
+        Constructs the circuit to solve the sudoku problem by implementing the
+        hadamard on the first 4 qubits, then the oracle gate across the
+        entire circuit, followed by the diffusion gate that acts on the first 4
+        qubits applying the oracle and diffuser twice to maximise the amplitude
+        of the solution
+
+        :return:
+            Matrix representing our completed Grover's algorithm for sudoku
+        """
+        had = g.multi_gate(9, [0, 1, 2, 3], g.Gate.H)
+        circuit = had
+
+        for i in range(2):
+            circuit = self.oracle() * circuit
+
+            circuit = self.diffusion() * circuit
+        return circuit
+
+    def run(self):
+        """
+        Multiplies our circuit with the initial state
+        :return: Column matrix representation of the final state
+        """
+        self.state = self.circuit * self.state
+        return self.state
+
+    def measure_state(self):
+        """
+        Randomly 'measures' self.state by selecting a state (out of 2**9)
+        weighted by its (amplitude ** 2)
+
+        returns:
+            Tuple[int, float]: The state observed and the probability of
+            measuring said state
+        """
+        p = reg.measure(self.state)
+        # list of weighted probabilities with the index representing the state
+
+        observed = random.choices(
+            [i for i in range(len(p))], p, k=1)  # type: ignore
+        probability = p[observed[0]]
+        return observed[0], probability
+
+    def measure_solution(self):
+        """
+        Randomly measures 1 of 16 possible options that the 4 input variables
+        can be with the probability of measuring a certain option determined by
+        the probabilities associated with each state
+        :return:
+            Tuple[[str], float]: The input solution observed and the
+            probability of observing that solution
+        """
+        prob = reg.measure(self.state)
+        sol_probs = [0] * 16
+        for i in range(len(prob)):
+            sol_probs[i % 16] += prob[i]
+
+        observed = random.choices(
+            [i for i in range(len(sol_probs))], sol_probs, k=1)  # type: ignore
+        chosen_prob = sol_probs[observed[0]]
+
+        vx = format(observed[0], '04b')
+        return [vx[-4], vx[-3], vx[-2], vx[-1]], chosen_prob
+
+
+def example():
```

Comment:
Just a question for when it comes to UI-ifying/CLI-ing this. Is there any scope for user input, or is it just run it and get the results?

Reply:
it just runs, I couldn't figure out a way to have active input